File: src/main/java/com/github/wolray/line/io/CsvWriter.java
package com.github.wolray.line.io;

import java.io.BufferedWriter;
import java.io.IOException;

/**
 * @author wolray
 */
public class CsvWriter<T> extends LineWriter<T> {
    private final ValuesJoiner<T> joiner;
    private final String sep;

    CsvWriter(ValuesJoiner<T> joiner, String sep) {
        super(joiner.toFormatter(sep));
        this.joiner = joiner;
        this.sep = sep;
    }

    @Override
    public Session write(Iterable<T> iterable) {
        return new Session(iterable);
    }

    public class Session extends LineWriter<T>.Session {
        private boolean utf8;

        protected Session(Iterable<T> iterable) {
            super(iterable);
        }

        public Session markUtf8() {
            utf8 = true;
            return this;
        }

        public Session withHeader() {
            addHeader(joiner.join(sep, a -> a.field.getName()));
            return this;
        }

        public Session columns(String... columns) {
            if (columns.length > 0) {
                addHeader(String.join(sep, columns));
            }
            return this;
        }

        @Override
        protected void preprocess(String file, BufferedWriter bw) throws IOException {
            if (utf8 && file.endsWith(".csv")) {
                bw.write('\ufeff');
            }
        }
    }
}

File: src/main/java/com/github/wolray/line/io/CsvWriter.kt
package com.github.wolray.line.io

import java.io.BufferedWriter

/**
 * @author wolray
 */
class CsvWriter<T> internal constructor(
    private val joiner: ValuesJoiner<T>,
    private val sep: String
) : LineWriter<T>(joiner.toFormatter(sep)) {

    override fun write(file: String) = Session(file)

    inner class Session internal constructor(file: String) : LineWriter<T>.Session(file) {
        private var utf8 = false

        override fun markUtf8() = apply { utf8 = true }
        override fun autoHeader() = apply { addHeader(joiner.joinFields(sep)) }

        override fun columnNames(vararg names: String) = apply {
            if (names.isNotEmpty()) {
                addHeader(names.joinToString(sep))
            }
        }

        override fun preprocess(bw: BufferedWriter) {
            if (utf8 && file.endsWith(".csv")) {
                bw.write('\ufeff'.code)
            }
        }
    }

    companion object {
        @JvmStatic
        fun <T> of(sep: String, type: Class<T>) = byCsv(sep, type)
    }
}
